CIP-0127? | Integration of `ripemd_160` into Plutus (#826)
cardano-foundation/CIPs: CIP-????/README.md

---

> CIP: ????

Reviewer: ```suggestion
CIP: 127
```

> Title: Integration of `ripemd-160` into Plutus

Reviewer 2: We can review naming on next Plutus crypto addition

Status: Proposed
Category: Plutus
Authors:
  - Tomasz Rybarczy <tomasz.rybarczyk@iohk.io>
Implementors: []
Discussions:
  - https://github.com/cardano-foundation/CIPs/pull/826
Created: 2024-05-22
License: Apache-2.0
---

## Abstract

> This CIP follows closely the (CIP-0101)[../CIP-0101/README.md][^1] and proposes an extension of the current Plutus functions with another hashing primitive [`RIPEMD-160`](https://en.bitcoin.it/wiki/RIPEMD-160). Primary goal is to introduce compatibility with Bitocoin's cryptographic infrastructure.

Reviewer 2: ```suggestion
This CIP follows closely the (CIP-0101)[^1] and proposes an extension of the current Plutus functions with another hashing primitive [`RIPEMD-160`](https://en.bitcoin.it/wiki/RIPEMD-160). Primary goal is to introduce compatibility with Bitcoin's cryptographic infrastructure.
```
Fix link and spelling

Author: Thank you!

Most of common addresses in Bitcoin are derived from double [hashing procedure](https://learnmeabitcoin.com/technical/cryptography/hash-function/#hash160) involving `SHA-256` followed by `RIPEMD-160` function:
- [P2KH](https://learnmeabitcoin.com/technical/script/p2pkh/)
- [P2WPKH](https://learnmeabitcoin.com/technical/script/p2wpkh/)
- [P2SH](https://learnmeabitcoin.com/technical/script/p2sh/)
- [P2WSH](https://learnmeabitcoin.com/technical/script/p2wsh/)

Adding `RIPEMD-160` to Plutus would enhance the potential for cross-chain solutions between Cardano and Bitcoin blockchains and complements the set of primitives which we already have in that regard. It would allow for the verification of Bitcoin addresses and transactions on-chain. This addition enables also the verification of signed messages that identify the signer by the public key hash, which has not yet been witnessed on the Bitcoin blockchain.

> The RIPEMD-160 is not only relevant to Bitcoin - other chains like (Cosmos)[https://docs.cosmos.network/main/build/architecture/adr-028-public-key-addresses#legacy-public-key-addresses-dont-change] or (BNB)[https://docs.bnbchain.org/docs/beaconchain/learn/accounts/#address] also use it for address generation.

Reviewer 2: ```suggestion
The RIPEMD-160 is not only relevant to Bitcoin - other chains like [Cosmos](https://docs.cosmos.network/main/build/architecture/adr-028-public-key-addresses#legacy-public-key-addresses-dont-change) or [BNB](https://docs.bnbchain.org/docs/beaconchain/learn/accounts/#address) also use it for address generation.
```

---

> CIP: 101

Reviewer: ```suggestion
CIP: ?
```
@paluh I can see you got permission to copy some of the text... but it's not helpful to copy this particular item 😅

Author: :-)

Title: Integration of `ripemd-160` into Plutus
Status: Proposed
Category: Plutus
Authors:
  - Tomasz Rybarczy <tomasz.rybarczyk@iohk.io>

> Implementors:

Reviewer: If memory serves, @michaelpj has recommended that implementors should be mentioned here by name (personally, not just the IOG Plutus Team as per the current Implementation Plan).

Reviewer 3: I no longer work for IOG, so I don't know whether the IOG Plutus Team has agreed to take this on, but yeah, someone needs to! I think CIPs with no implementers just tend to disappoint people.

Reviewer: @zliu41 I guess the first question for `Implementors:` is to establish whether IOG has this curve inclusion on their agenda as we first asked in https://github.com/cardano-foundation/CIPs/pull/826#pullrequestreview-2074631093?

Author: @rphair  so I'm happy to help with the implementation but I'm not able to cover it all so for sure some people will be added to the list as we go.

Reviewer: OK @paluh then the standard form would be `Implementors: []` (the empty set) for now & I'm adding this change request to the comment above.

> Discussions:

Reviewer: ```suggestion
Discussions:
  - https://github.com/cardano-foundation/CIPs/pull/826
```

More specifically, Plutus will gain the following primitive operation:

* `ripemd_160` :: ByteString -> ByteString

The input to this function can be a `ByteString` of arbitrary size, and the output will be a `ByteString` of 20 bytes. 
Note that this function aligns with the format of existing hash functions in Plutus, such as [blake2b_256](https://github.com/input-output-hk/plutus/blob/75267027f157f1312964e7126280920d1245c52d/plutus-core/plutus-core/src/Data/ByteString/Hash.hs#L25)

## Rationale: how does this CIP achieve its goals?
While the `RIPEMD-160` function might be implemented in on-chain scripts, doing so would be computationally unfeasible.

The library, cryptonite, is not implemented by and under control of the Plutus team. However,
* It is a library already used in the Plutus stack to expose KECCAK-256, and can be considered as a trustworthy implementation.
* Its behaviour is predictable and computationally efficient. The cost of the function is linear with respect to the size of the message provided as input. This is the same behaviour that other hash functions exposed in plutus (blake, sha3, keccak-256) have.

## Path to Active

> This CIP may transition to active status once the Plutus version containing the `ripemd_160` function is introduced in a node release and becomes available on Mainnet.

Reviewer: Since this point refers to granting `Active` status, it should be moved into the list of _Acceptance Criteria_.
